cms/djangoapps/contentstore/views/tests/test_item.py

```diff
@@ -680,6 +681,152 @@ def verify_name(source_usage_key, parent_usage_key, expected_name, display_name=
+        self.assertEqual(response['source_index'], source_index)
+        new_parent_loc = self.store.get_parent_location(source_usage_key)
+        self.assertEqual(new_parent_loc, dest_usage_key)
+        self.assertNotEqual(parent_loc, new_parent_loc)
+
+    def test_move_component(self):
+        """
+        Test move component with different xblock types.
+        """
+        for source_usage_key, dest_usage_key in [
+                (self.html_usage_key, self.vert2_usage_key),
+                (self.vert_usage_key, self.seq2_usage_key),
+                (self.seq_usage_key, self.chapter2_usage_key)
+        ]:
+            self.assert_move_item(source_usage_key, dest_usage_key)
+
+    def test_move_source_index(self):
+        """
+        Test moving an item to a particular index.
+        """
+        parent = self.get_item_from_modulestore(self.vert_usage_key)
+        children = parent.get_children()
+        self.assertEqual(len(children), 2)
+
+        # Create a component within vert2.
+        resp = self.create_xblock(parent_usage_key=self.vert2_usage_key, display_name='html2', category='html')
+        html2_usage_key = self.response_usage_key(resp)
+
+        # Move html2_usage_key inside vert_usage_key at second position.
+        self.assert_move_item(html2_usage_key, self.vert_usage_key, 1)
+        parent = self.get_item_from_modulestore(self.vert_usage_key)
+        children = parent.get_children()
+        self.assertEqual(len(children), 3)
+        self.assertEqual(children[1].location, html2_usage_key)
+
+    def test_move_undo(self):
+        """
+        Test move a component and move it back (undo).
+        """
+        self.assert_move_item(self.html_usage_key, self.vert2_usage_key)
+        self.assert_move_item(self.html_usage_key, self.vert_usage_key)
+
+    def test_move_moved_parent(self):
+        """
+        Test move a moved parents.
+        """
+        self.assert_move_item(self.vert_usage_key, self.seq2_usage_key)
+        self.assert_move_item(self.seq2_usage_key, self.chapter2_usage_key)
+
+        # check vert_usage_key has seq2_usage_key as parent
+        self.assertEqual(self.store.get_parent_location(self.vert_usage_key), self.seq2_usage_key)
+
+    def test_invalid_move(self):
+        """
+        Test invalid move.
+        """
+        parent_loc = self.store.get_parent_location(self.chapter_usage_key)
+        response = self._move_component(self.chapter_usage_key, self.usage_key)
+        self.assertEqual(response.status_code, 400)
+        response = json.loads(response.content)
+
+        expected_error = 'You can not move {source_type} into {dest_type}'.format(
+            source_type=self.chapter_usage_key.block_type,
+            dest_type=self.usage_key.block_type
+        )
+        self.assertEqual(expected_error, response['error'])
+        new_parent_loc = self.store.get_parent_location(self.chapter_usage_key)
+        self.assertEqual(new_parent_loc, parent_loc)
+
```

Comment:
These are great tests. I'd add a few more:

 - verify what happens if the user doesn't have write permission for the source and/or target
 - verify that the change is transactional if one part fails
 - verify what happens if you move to an invalid index
 - verify what happens if you move a block to its current parent (i.e. a no-op)
 - verify what happens if the parent locator does not exist